math-scala/src/main/scala/org/apache/mahout/math/algorithms/clustering/rtree/Rectangle.scala

```diff
@@ -0,0 +1,157 @@
+        false
+      }
+    }
+    true
+  }
+
+  def wraps(rect: Rectangle): Boolean = {
+    for(i <- 0 until rect.n ) {
+      if( !(this.bottomLeft(i) < rect.bottomLeft(i) && rect.topRight(i) < this.topRight(i)) ) {
+        false
+      }
+    }
+    true
+  }
+
+  def intersects(rect: Rectangle): Boolean = {
+    for(i <- 0 until rect.n) {
+      if( !(this.bottomLeft(i) <= rect.topRight(i) && rect.bottomLeft(i) <= this.topRight(i)) )
+        false
+    }
+    true
+  }
+
+  def area(): Double = {
+    var area = 1.0
+    for(i <- 0 until this.n) {
+      area = area * (this.topRight(i) - this.bottomLeft(i))
+    }
+    area
+  }
+
+  def toMBR: MBR = MBR(this.bottomLeft, this.topRight)
+
+  def expandRectangle(rect: Rectangle): MBR = {
+    val minBottomLeft = new Array[Double](rect.n)
+    val maxTopRight = new Array[Double](rect.n)
+    for(i <- 0 until rect.n) {
+      minBottomLeft(i) = math.min(bottomLeft(i), rect.bottomLeft(i))
+      maxTopRight(i) = math.max(topRight(i), rect.topRight(i))
+    }
+    MBR(minBottomLeft.toList, maxTopRight.toList)
+  }
+
+  //Area of 'this' rectangle after merging with 'rect'
+  def diffAreaAfterExpansion(rect: Rectangle): Double = {
+    val minBottomLeft = new Array[Double](rect.n)
+    val maxTopRight = new Array[Double](rect.n)
+    var area: Double = 0
+
+    for(i <- 0 until rect.n) {
+      minBottomLeft(i) = math.min(bottomLeft(i), rect.bottomLeft(i))
+      maxTopRight(i) = math.max(topRight(i), rect.topRight(i))
+      area += (maxTopRight(i) - minBottomLeft(i))*(maxTopRight(i) - minBottomLeft(i))
+    }
+    area - this.area
+  }
+
+  //Distance of a point from the rectangle.
+  def distance(pt: Point): Double = {
+    val n = pt.coordinates.size
+    val dList = new Array[Double](n)
+    for(dim <- 0 until n) {
+      var x = this.bottomLeft(dim)
+      var x2 = this.topRight(dim)
+      dList(dim) = if (pt.coordinates(dim) < x) x - pt.coordinates(dim) else if (pt.coordinates(dim) < x2) 0D else pt.coordinates(dim) - x2
+    }
+    var squaredSum: Double = 0
+    for(i <- 0 until n) {
+      squaredSum += dList(i)*dList(i)
+    }
+    sqrt(squaredSum)
+  }
+
+}
+
+//Represents a point with coordinates and n information
+case class Point(coordinates: List[Double]) extends Rectangle {
+
+  override def bottomLeft: List[Double] = coordinates
+  override def topRight: List[Double] = coordinates
+
+  override def n: Int = bottomLeft.size
+
+  override def lowerLeftPoint: Point = this
+  override def upperRightPoint: Point = this
+
+  override def distance(pt: Point): Double = {
+    var squaredSum: Double = 0
+    for(i <- 0 until pt.n){
+      squaredSum += (pt.bottomLeft(i) - bottomLeft(i))*(pt.bottomLeft(i) - bottomLeft(i))
+    }
+    squaredSum
+  }
+
+  override def wraps(rect: Rectangle): Boolean = false
+}
+
+//MBR is a physical manifestation of the Rectangle trait
+case class MBR(bottomLeft: List[Double], topRight: List[Double]) extends Rectangle {
```

Comment:
same with case class and companion object. (move to own file)

Reply:
My idea was to keep classes describing the datastructures in one file and the logic in another. But if having different classes is food design I'll update it